pkg/controllers/updaterun/initialization.go

```diff
@@ -0,0 +1,433 @@
+	}
+	if err := r.Client.List(ctx, &policySnapshotList, latestPolicyMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list the latest policy snapshots", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		// err can be retried.
+		return nil, -1, err
+	}
+	if len(policySnapshotList.Items) != 1 {
+		if len(policySnapshotList.Items) > 1 {
+			err := fmt.Errorf("more than one (%d in actual) latest policy snapshots associated with the clusterResourcePlacement: %s", len(policySnapshotList.Items), placementName)
+			klog.ErrorS(controller.NewUnexpectedBehaviorError(err), "Failed to find the latest policy snapshot", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+			// no more retries for this error.
+			return nil, -1, fmt.Errorf("%w: %s", errInitializedFailed, err.Error())
+		}
+		// no latest policy snapshot found.
+		err := fmt.Errorf("no latest policy snapshot associated with the clusterResourcePlacement: %s", placementName)
+		klog.ErrorS(err, "Failed to find the latest policy snapshot", "clusterResourcePlacement", placementName, "clusterStagedUpdateRun", updateRunRef)
+		// again, no more retries here.
+		return nil, -1, fmt.Errorf("%w: %s", errInitializedFailed, err.Error())
+	}
+
+	latestPolicySnapshot := policySnapshotList.Items[0]
+	updateRun.Status.PolicySnapshotIndexUsed = latestPolicySnapshot.Name
+	// Get the cluster count from the policy snapshot.
+	clusterCount, err := annotations.ExtractNumOfClustersFromPolicySnapshot(&latestPolicySnapshot)
+	if err != nil {
+		annErr := fmt.Errorf("%w: the policySnapshot `%s` doesn't have valid cluster count annotation", err, latestPolicySnapshot.Name)
+		klog.ErrorS(controller.NewUnexpectedBehaviorError(annErr), "Failed to get the cluster count from the latestPolicySnapshot", "clusterResourcePlacement", placementName, "latestPolicySnapshot", latestPolicySnapshot.Name, "clusterStagedUpdateRun", updateRunRef)
+		// no more retries here.
+		return nil, -1, fmt.Errorf("%w, %s", errInitializedFailed, annErr.Error())
+	}
+	updateRun.Status.PolicyObservedClusterCount = clusterCount
+	klog.V(2).InfoS("Found the latest policy snapshot", "latestPolicySnapshot", latestPolicySnapshot.Name, "observed cluster count", updateRun.Status.PolicyObservedClusterCount, "clusterStagedUpdateRun", updateRunRef)
+
+	if !condition.IsConditionStatusTrue(latestPolicySnapshot.GetCondition(string(placementv1beta1.PolicySnapshotScheduled)), latestPolicySnapshot.Generation) {
+		scheduleErr := fmt.Errorf("policy snapshot `%s` not fully scheduled yet", latestPolicySnapshot.Name)
+		klog.ErrorS(scheduleErr, "The policy snapshot is not scheduled successfully", "clusterResourcePlacement", placementName, "latestPolicySnapshot", latestPolicySnapshot.Name, "clusterStagedUpdateRun", updateRunRef)
+		// hmmmm, should we retry and see if the policy snapshot is scheduled later?
+		return nil, -1, fmt.Errorf("%w: %s", errInitializedFailed, scheduleErr.Error())
```

Comment:
Should we retry here instead of reporting initialization failure?

Reply:
This function will be reused by the validation part where it should fail. I think the expectation is that the updateRun is only issued after we have a stable schedule.